### ISSUE TITLE
Stable/Predictable  Lifecycle hooks for extensions (NOT activates/deactivates)

### ISSUE THREAD

**Issue by Jaimeloeuf** [microsoft/vscode#98732]
## This is a feature request for LifeCycle hooks API for extensions
Based on my research this has been requested / talked about in the past, but to no avail or concrete updates yet. Here are some conversations around this topic:
- [Visual Studio Code: Installation hook for snippets extension](https://stackoverflow.com/questions/42812416/visual-studio-code-installation-hook-for-snippets-extension)
- [Extension deactivate() doesn't call #47881](https://github.com/Microsoft/vscode/issues/47881) where deactivate does not behave as one might expect
- [Provide access to VS Code API in extension uninstall lifecycle hook #45474](https://github.com/microsoft/vscode/issues/45474)
- There are not many questions regarding this in [stack overflow](https://stackoverflow.com/search?q=%5Bvscode-extensions%5D+lifecycle+hook) too
- I am not sure if [this](https://github.com/codota/tabnine-vscode/pull/97) is related but will just add it in here

## Feature requested
Lifecycle hooks similar to those from Vue JS / React JS, where the extensions would be able to define functions to run for the lifecycle events.
Although hooks like those from vue would be great, those from chrome extensions seem usable too albeit not as extensive or fine grained
- <https://stackoverflow.com/a/40614389/13137262>
- <https://developer.chrome.com/apps/app_lifecycle>
- <https://developer.chrome.com/extensions/runtime>

Here are some suggested hooks if this feature becomes a thing:
- onInstall
- onUpdate
- onUninstall
- onEnable
- onDisable

## Purpose of these hooks and how they can be used
These hooks can be used for a wide variety of stuff, here are just some basic examples
- running 1 time setup processes after installation
- run configuration change or what not when the user updates the extension
- telemetry and usage reporting
- clean ups during uninstallation
- ... and many many more

## How is this different from activation / deactivation methods
- Because activation runs every single time you use/activate the extension using any one of the activation events, which means that if you place any setup code here, they will be ran repeatedly over and over again when vs code relaunches. Especially true when u are using the "*" activation event.
- Thus this is not really similar to that of a onInstall hook.
- For example, if I want to run a setup function that only runs post installation to update user settings, I can with code in activation method, but it will be ran repeatedly and not only once.
- This means reduced efficiency and longer startup/load times
    - I suspect this might be why some extensions like eslint are slow when vs code first opens / during the first save
    - They might probably be running their setup codes on activation every single time before registering and executing the commands. Or they can only setup event listeners and act upon those events on activation which makes them slow.
- I am not suggesting for activation event to be dropped in favour of lifecycle hooks, but having more fine grained extension lifecycle hooks can make for a better vs code and extensions thanks to more efficient operations.

## Example
- This [file](https://github.com/Enkel-Digital/relative-goto/blob/1da3dd4332ab6de296a37d12e1cc2fe1684f5a64/src/extension.ts#L18
) calls this [module](https://github.com/Enkel-Digital/relative-goto/blob/1da3dd4332ab6de296a37d12e1cc2fe1684f5a64/src/setup.ts) everytime before registering the commands, which causes 
    1. a visible lag time before the command can be executed
    2. inefficiency
    3. potential bugs where you only want it to be ran once
- For example, right now for a simple extension with my current setup function, takes about 3-6 seconds to activate and allow first run of command, and below is my system specs
    - cpu: i5-8250u
    - ram: 16GB 
    - disk: nvme SSD (not sure about exact specs but basically should be super fast)

## Side note
- When I was first starting out, I was confused and thought that activation is only for first install which makes more sense then actiavating everytime before u use it, from a beginner point of view
- We should update docs to make the definition of activation clearer, as it is super easy to confuse it as a setup / onInstall hook since the generated template with ```yo code``` says that "These code will only be executed once when your extension is activated" which can easily be interpreted as installation hook if activation is not clearly defined.
- Right now, I am not sure how to proceed without directly digging into the vscode Core to see if this is possible, but would love it if someone familiar with this part of the architecture can provide some insights for this.
- If I'm missing out something do let me know! Pretty new to this, have only been working with vs code extensions for 3 days 😅

**Comment by anburocky3** [comment]
Any update on this?

**Comment by VSCodeTriageBot** [comment]
We closed this issue because we [don't plan to address it](https://aka.ms/vscode-out-of-scope) in the foreseeable future. If you disagree and feel that this issue is crucial: we are happy to listen and to reconsider.

If you wonder what we are up to, please see our [roadmap](https://aka.ms/vscoderoadmap) and [issue reporting guidelines](https://aka.ms/vscodeissuereporting).

Thanks for your understanding, and happy coding!
